toolkit/primitives/block-rewards-payouts/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+		fn blocks_to_reward(slot: Slot) -> Vec<(Slot, BlockProducerId)>;
+	}
+}
+
+#[cfg(feature = "std")]
+pub mod inherent_data {
+	use super::*;
+	use alloc::fmt::Debug;
+	use core::error::Error;
+	use sidechain_domain::mainchain_epoch::*;
+	use sidechain_domain::*;
+	use sp_api::{ApiError, ProvideRuntimeApi};
+	use sp_inherents::{InherentData, InherentDataProvider};
+	use sp_runtime::traits::Block as BlockT;
+	pub use sp_stake_distribution::StakeDistributionDataSource;
+	use std::collections::HashMap;
+	use std::hash::Hash;
+
+	#[derive(thiserror::Error, Debug)]
+	pub enum InherentDataCreationError {
+		#[error("Runtime API call failed: {0}")]
+		ApiError(#[from] ApiError),
+		#[error("Data source call failed: {0}")]
+		DataSourceError(Box<dyn Error + Send + Sync>),
+	}
+
+	#[derive(Debug)]
+	pub struct BlockRewardsPayoutsInherentDataProvider<BlockProducerId, BeneficiaryId> {
+		block_production_data: Option<BlockProductionPayoutInfo<BlockProducerId, BeneficiaryId>>,
+	}
+
+	impl<BlockProducerId, BeneficiaryId>
+		BlockRewardsPayoutsInherentDataProvider<BlockProducerId, BeneficiaryId>
+	{
+		pub async fn new_cardano_stake_based<Block: BlockT, T>(
+			client: &T,
+			stake_distribution_data_source: &(dyn StakeDistributionDataSource + Send + Sync),
+			parent_hash: <Block as BlockT>::Hash,
+			current_slot: Slot,
+			mc_epoch_config: &MainchainEpochConfig,
+			slot_duration: SlotDuration,
+		) -> Result<Self, InherentDataCreationError>
+		where
+			T: ProvideRuntimeApi<Block> + Send + Sync,
+			T::Api: BlockRewardsPayoutsApi<Block, BlockProducerId>,
+			BlockProducerId: Decode + Clone + Hash + Eq + Ord + Debug,
+			BeneficiaryId: From<DelegatorKey> + Ord + Debug,
+			Option<MainchainKeyHash>: From<BlockProducerId>,
+		{
+			let api = client.runtime_api();
+			let Some(up_to_slot) = api.should_pay_out(parent_hash, current_slot)? else {
+				log::debug!("💤︎ Skipping block rewards payouts this block...");
+				return Ok(Self { block_production_data: None });
+			};
+			let blocks_to_reward = api.blocks_to_reward(parent_hash, up_to_slot)?;
+
+			let mut block_production_payout_info =
+				BlockProductionPayoutInfo { up_to_slot, production_rundowns: vec![] };
+
+			let epoch_map =
+				count_by_epoch_and_producer(blocks_to_reward, mc_epoch_config, slot_duration);
+
+			for (mc_epoch, producers) in epoch_map.into_iter() {
+				let pools: Vec<_> =
+					producers.keys().cloned().flat_map(Into::<Option<_>>::into).collect();
+				let distribution = stake_distribution_data_source
+					.get_stake_pool_delegation_distribution_for_pools(mc_epoch, &pools)
+					.await
+					.map_err(InherentDataCreationError::DataSourceError)?;
+				for (producer, block_count) in producers.into_iter() {
+					let (beneficiary_total_share, beneficiaries) = match producer.clone().into() {
+						None => (0, vec![]),
+						Some(cardano_producer) => {
+							let PoolDelegation { total_stake, delegators } =
+								distribution.0.get(&cardano_producer).expect("");
```

Comment:
This assumes that there's always a value - maybe we want to be more careful to get a useful error instead of a panic..

```Rust
let PoolDelegation { total_stake, delegators } =
    distribution.0.get(&cardano_producer)
    .ok_or_else(|| InherentDataCreationError::DataSourceError("Missing delegation data".into()))?;
```